# 0. Import our libraries

In [1]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns

# 1. Collecting data
- What subject is your data about? What is the source of your data?  
- Do authors of this data allow you to use like this? You can check the data license  
- How did authors collect data?


## About dataset
### Subject: Commercial flight operations in the United States (2024)

Although air travel is essential for bringing people and economies together, passenger flights are frequently disrupted by delays and cancellations. With over 1 million flights nationwide, this dataset offers a thorough understanding of US airline performance in 2024.

The dataset, which includes comprehensive statistics on departure times, cancellations, flight distances, weather delays, and late aircraft delays, presents a special chance to investigate:

### Sources
The On-Time Performance Data (2024) from the US Bureau of Transportation Statistics (BTS) is the source of this dataset. The source offers comprehensive details on airline on-time performance, including departure times, weather-related disruptions, delays, and cancellations.

## Data Lisence

**No Copyright** 

The person who associated a work with this deed has dedicated the work to the public domain by waiving all of his or her rights to the work worldwide under copyright law, including all related and neighboring rights, to the extent allowed by law. 

You can copy, modify, distribute and perform the work, even for commercial purposes, all without asking permission. See Other Information via the link below. 

[CC0: Public Domain](https://creativecommons.org/publicdomain/zero/1.0/)



## Collection Methodology

The information was gathered from the 2024 On-Time Performance reports published by the US Bureau of Transportation Statistics (BTS). A single CSV file containing important features, such as delays, cancellations, distances, and weather impacts, was created by merging, cleaning, and curating monthly reports.

# 2. Exploring data 
- How many rows and how many columns?  
- What is the meaning of each row?  
- Are there **<span style="color:red">duplicated rows</span>**?  
- What is the meaning of each column?  
- What is the current data type of each column? Are there columns having **<span style="color:red">inappropriate data types</span>**?  
- With each numerical column, how are values distributed?  
  - What is the percentage of **<span style="color:red">missing values</span>**?  
  - Min? max? Are they **<span style="color:red">abnormal</span>**?  
- With each categorical column, how are values distributed?  
  - What is the percentage of **<span style="color:red">missing values</span>**?  
  - How many different values? Show a few  
  - Are they **<span style="color:red">abnormal</span>**?

## How many rows and columns?

Get to know our dataframe first!!

In [6]:
df = pd.read_csv("data/flight_data_2024.csv")
df.head(10)

year  month  day_of_month  day_of_week   fl_date origin  \
0  2024      1             1            1  1/1/2024    JFK   
1  2024      1             1            1  1/1/2024    MSP   
2  2024      1             1            1  1/1/2024    JFK   
3  2024      1             1            1  1/1/2024    RIC   
4  2024      1             1            1  1/1/2024    DTW   
5  2024      1             1            1  1/1/2024    JAX   
6  2024      1             1            1  1/1/2024    LGA   
7  2024      1             1            1  1/1/2024    CHS   
8  2024      1             1            1  1/1/2024    LGA   
9  2024      1             1            1  1/1/2024    ITH   

      origin_city_name origin_state_nm  dep_time  taxi_out  wheels_off  \
0         New York, NY        New York    1247.0      31.0      1318.0   
1      Minneapolis, MN       Minnesota    1001.0      20.0      1021.0   
2         New York, NY        New York    1411.0      21.0      1432.0   
3         Richmond, VA        Virginia    1643.0      13.0      1656.0   
4          Detroit, MI        Michigan    1010.0      21.0      1031.0   
5     Jacksonville, FL         Florida    1403.0      14.0      1417.0   
6         New York, NY        New York     947.0      26.0      1013.0   
7       Charleston, SC  South Carolina    1135.0       8.0      1143.0   
8         New York, NY        New York     810.0      14.0       824.0   
9  Ithaca/Cortland, NY        New York    1248.0      12.0      1300.0   

   wheels_on  taxi_in  cancelled  air_time  distance  weather_delay  \
0     1442.0      7.0          0      84.0       509              0   
1     1249.0      6.0          0      88.0       622              0   
2     1533.0      8.0          0      61.0       288              0   
3     1747.0     12.0          0      51.0       288              0   
4     1016.0      4.0          0      45.0       237              0   
5     1559.0      4.0          0     102.0       833              0   
6     1218.0     13.0          0     125.0       833              0   
7     1309.0      5.0          0      86.0       641              0   
8     1005.0      8.0          0     101.0       641              0   
9     1343.0     12.0          0      43.0       189              0   

   late_aircraft_delay  
0                    0  
1                    0  
2                    0  
3                    0  
4                    0  
5                    0  
6                    0  
7                    0  
8                    0  
9                    0

In [5]:
df.tail()

year  month  day_of_month  day_of_week    fl_date origin  \
1048570  2024      2            29            4  2/29/2024    SJU   
1048571  2024      2            29            4  2/29/2024    PHX   
1048572  2024      2            29            4  2/29/2024    JAX   
1048573  2024      2            29            4  2/29/2024    CLT   
1048574  2024      2            29            4  2/29/2024    MCI   

         origin_city_name origin_state_nm  dep_time  taxi_out  wheels_off  \
1048570      San Juan, PR     Puerto Rico    1447.0      24.0      1511.0   
1048571       Phoenix, AZ         Arizona      59.0      12.0       111.0   
1048572  Jacksonville, FL         Florida    2014.0      20.0      2034.0   
1048573     Charlotte, NC  North Carolina    1104.0      12.0      1116.0   
1048574   Kansas City, MO        Missouri    1307.0      13.0      1320.0   

         wheels_on  taxi_in  cancelled  air_time  distance  weather_delay  \
1048570     1640.0      7.0          0     149.0      1045              0   
1048571      647.0      3.0          0     216.0      1972              0   
1048572     2128.0     10.0          0      54.0       328              0   
1048573     1210.0      7.0          0     114.0       808              0   
1048574     1559.0     22.0          0      99.0       808              0   

         late_aircraft_delay  
1048570                    0  
1048571                  105  
1048572                    0  
1048573                   70  
1048574                  104

In [3]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1048575 entries, 0 to 1048574
Data columns (total 18 columns):
 #   Column               Non-Null Count    Dtype  
---  ------               --------------    -----  
 0   year                 1048575 non-null  int64  
 1   month                1048575 non-null  int64  
 2   day_of_month         1048575 non-null  int64  
 3   day_of_week          1048575 non-null  int64  
 4   fl_date              1048575 non-null  object 
 5   origin               1048575 non-null  object 
 6   origin_city_name     1048575 non-null  object 
 7   origin_state_nm      1048575 non-null  object 
 8   dep_time             1026022 non-null  float64
 9   taxi_out             1025450 non-null  float64
 10  wheels_off           1025450 non-null  float64
 11  wheels_on            1024898 non-null  float64
 12  taxi_in              1024898 non-null  float64
 13  cancelled            1048575 non-null  int64  
 14  air_time             1022824 non-null  float64
 15

We can easily see that:
- Number of rows: 1,048,575
- Number of columns: 18


## The meaning of each row and column?
### The meaning of each row
- Each row corresponds to a single flight (one origin to one destination) in the time‐span covered by the dataset.
### The meaning of each column
| **Column name**         | **Meaning**                                                                                                           |
| ----------------------- | --------------------------------------------------------------------------------------------------------------------- |
| **year**                | The year when the flight occurred (e.g., 2024).                                                                       |
| **month**               | The month number (1 = January, 12 = December).                                                                        |
| **day_of_month**        | The day of the month (1–31).                                                                                          |
| **day_of_week**         | The weekday number (1 = Monday, 7 = Sunday).                                                                          |
| **fl_date**             | The full flight date (month/day/year).                                                                                |
| **origin**              | The **IATA code** of the origin airport (e.g., JFK = John F. Kennedy International Airport).                          |
| **origin_city_name**    | The full city and state of the origin airport (e.g., “New York, NY”).                                                 |
| **origin_state_nm**     | The full name of the U.S. state where the origin airport is located (e.g., “New York”).                               |
| **dep_time**            | The **actual departure time** in local time, usually given in HHMM format (e.g., 1247 → 12:47 PM).                    |
| **taxi_out**            | The number of minutes the aircraft spent **taxiing on the ground before takeoff** (from gate pushback to wheels-off). |
| **wheels_off**          | The actual **time when the aircraft lifted off the runway**, in HHMM local time.                                      |
| **wheels_on**           | The actual **time when the aircraft touched down at the destination**, in HHMM local time.                            |
| **taxi_in**             | The number of minutes spent **taxiing after landing** (from touchdown to gate arrival).                               |
| **cancelled**           | A binary indicator: **0 = not cancelled**, **1 = cancelled**.                                                         |
| **air_time**            | The **total time in the air** (in minutes), equal to `wheels_on - wheels_off` (converted).                            |
| **distance**            | The great-circle **distance between origin and destination** (in miles).                                              |
| **weather_delay**       | Minutes of delay caused by **weather-related issues** (0 if none).                                                    |
| **late_aircraft_delay** | Minutes of delay caused by the **late arrival of the same aircraft** from a previous flight (0 if none).              |


In [8]:
df.describe()

year         month  day_of_month   day_of_week      dep_time  \
count  1048575.0  1.048575e+06  1.048575e+06  1.048575e+06  1.026022e+06   
mean      2024.0  1.478081e+00  1.530512e+01  3.893483e+00  1.325074e+03   
std          0.0  4.995196e-01  8.585503e+00  2.010038e+00  4.972990e+02   
min       2024.0  1.000000e+00  1.000000e+00  1.000000e+00  1.000000e+00   
25%       2024.0  1.000000e+00  8.000000e+00  2.000000e+00  9.110000e+02   
50%       2024.0  1.000000e+00  1.500000e+01  4.000000e+00  1.323000e+03   
75%       2024.0  2.000000e+00  2.300000e+01  6.000000e+00  1.736000e+03   
max       2024.0  2.000000e+00  3.100000e+01  7.000000e+00  2.400000e+03   

           taxi_out    wheels_off     wheels_on       taxi_in     cancelled  \
count  1.025450e+06  1.025450e+06  1.024898e+06  1.024898e+06  1.048575e+06   
mean   1.825012e+01  1.349996e+03  1.476156e+03  8.082517e+00  2.222635e-02   
std    1.044025e+01  4.980426e+02  5.198682e+02  6.512591e+00  1.474190e-01   
min    1.000000e+00  1.000000e+00  1.000000e+00  1.000000e+00  0.000000e+00   
25%    1.200000e+01  9.290000e+02  1.058000e+03  4.000000e+00  0.000000e+00   
50%    1.500000e+01  1.337000e+03  1.510000e+03  6.000000e+00  0.000000e+00   
75%    2.100000e+01  1.750000e+03  1.914000e+03  9.000000e+00  0.000000e+00   
max    2.130000e+02  2.400000e+03  2.400000e+03  4.440000e+02  1.000000e+00   

           air_time      distance  weather_delay  late_aircraft_delay  
count  1.022824e+06  1.048575e+06   1.048575e+06         1.048575e+06  
mean   1.162270e+02  8.345389e+02   1.194321e+00         5.326660e+00  
std    7.091204e+01  5.923104e+02   2.005819e+01         2.975676e+01  
min    8.000000e+00  3.100000e+01   0.000000e+00         0.000000e+00  
25%    6.400000e+01  4.020000e+02   0.000000e+00         0.000000e+00  
50%    1.000000e+02  6.920000e+02   0.000000e+00         0.000000e+00  
75%    1.470000e+02  1.069000e+03   0.000000e+00         0.000000e+00  
max    7.230000e+02  5.095000e+03   1.804000e+03         2.100000e+03

# 3. Asking meaningful questions

- Your group needs to give ≥ [the-number-of-group-members](#) questions which can be answered with this data.  
- Each question should be [meaningful](#) (what are benefits of finding the answer?) and [not too easy](#) to answer (e.g., it’s too easy if we just need one line of code to get the answer).  
- Your group should focus more on the [quality of questions](#) than the quantity.  
- In notebook file, with each question, your group needs to present:  
  - What is the question?  
  - What are benefits of finding the answer?


# 4. Preprocessing and analyzing data to answer each question

- With each question:  
  - Does it need to have preprocessing step, and if yes, how does your group preprocess?  
    - **Text:** sketch steps [clearly](#) so that readers can understand how your group preprocesses even without reading code  
    - **Code:** implement [sketched](#) steps. Your group should also try to write code clearly (choose good variable names, comment where should be commented, don’t let a line too long)  
  - How does your group analyze data to answer the question?  
    - **Text:** similar to above  
    - **Code:** similar to above


# 5. Reflection
- **Each member:** What difficulties have you encountered?  
- **Each member:** What have you learned?  
- **Your group:** If you had more time, what would you do?


# 6. References
- To finish this project, what materials have you consulted

# 7. Some general points:

- In notebook file, one important thing your group should try to practice is to **[organize sections](#)** and **[write/code clearly](#)**.  
  Your group should use Markdown headings to organize sections, and use Jupyter Notebook/Lab TOC to quickly navigate through headings (it’s similar to bookmark in pdf file).  
  During the process of writing/coding, try to maintain a calm mind, try to **[think for readers](#)**.

- In data science process your group needs to do, I think the most time-consuming part is **[from finding data to giving meaningful questions which can be answered with data](#)**.

